Source/WebKit/UIProcess/API/Swift/WebPage.swift

```diff
@@ -120,20 +125,125 @@ public class WebPage_v0 {
+        return navigation.map(NavigationID.init(_:))
+    }
+
+    @discardableResult
+    public func loadSimulatedRequest(_ request: URLRequest, response: URLResponse, responseData: Data) -> NavigationID? {
+        // `WKWebView` annotates this method as returning non-nil, but it may return nil.
+
+        let navigation = backingWebView.loadSimulatedRequest(request, response: response, responseData: responseData) as WKNavigation?
+        return navigation.map(NavigationID.init(_:))
+    }
+
+    @discardableResult
+    public func loadSimulatedRequest(_ request: URLRequest, responseHTML: String) -> NavigationID? {
+        // `WKWebView` annotates this method as returning non-nil, but it may return nil.
+
+        let navigation = backingWebView.loadSimulatedRequest(request, responseHTML: responseHTML) as WKNavigation?
+        return navigation.map(NavigationID.init(_:))
+    }
+
+    @discardableResult
+    public func reload(fromOrigin: Bool = false) -> NavigationID? {
+        let navigation = fromOrigin ? backingWebView.reloadFromOrigin() : backingWebView.reload()
+        return navigation.map(NavigationID.init(_:))
+    }
+
+    public func stopLoading() {
+        backingWebView.stopLoading()
+    }
+
+    public func callAsyncJavaScript(_ functionBody: String, arguments: [String : Any] = [:], in frame: FrameInfo? = nil, contentWorld: WKContentWorld? = nil) async throws -> Any? {
+        try await backingWebView.callAsyncJavaScript(functionBody, arguments: arguments, in: frame?.wrapped, contentWorld: contentWorld ?? .defaultClient)
+    }
+
+#if canImport(UIKit)
+    public func snapshot(configuration: WKSnapshotConfiguration? = nil) async throws -> UIImage {
+        try await withCheckedThrowingContinuation { continuation in
+            backingWebView.takeSnapshot(with: configuration) { image, error in
+                if let error {
+                    continuation.resume(throwing: error)
+                    return
+                }
+
+                guard let image else {
+                    preconditionFailure()
+                }
+
+                continuation.resume(returning: image)
+            }
+        }
+    }
+#else
+    public func snapshot(configuration: WKSnapshotConfiguration? = nil) async throws -> NSImage {
+        try await withCheckedThrowingContinuation { continuation in
+            backingWebView.takeSnapshot(with: configuration) { image, error in
+                if let error {
+                    continuation.resume(throwing: error)
+                    return
+                }
+
+                guard let image else {
+                    preconditionFailure()
+                }
+
+                continuation.resume(returning: image)
+            }
+        }
+    }
+#endif
+
+    public func pdf(configuration: WKPDFConfiguration? = nil) async throws -> Data {
+        try await backingWebView.pdf(configuration: configuration ?? .init())
```

Comment:
I was almost going to suggest making the configuration non-optional above, but since `nil` would flow through all the layers it seemed unnecessary. This give some some pause since there really isn’t any reason to make that `= .init()` just the default value (it won’t incur any runtime penalty, default values are inlined into the client binary).

That does make me kinda wish we’d have consistent optionality, but I think even without this should just use a default value.